module/synchronization/core.go

```diff
@@ -237,13 +298,9 @@ func (c *Core) prune(final *flow.Header) {
 	}
 
 	for blockID, status := range c.blockIDs {
-		if status.WasReceived() {
-			header := status.Header
-
-			if header.Height <= final.Height {
-				delete(c.blockIDs, blockID)
-				continue
-			}
+		if status.WasReceived() && status.height <= final.Height {
+			delete(c.blockIDs, blockID)
+			continue
```

Comment:
The pruning could exit earlier if we iterate from low to high through a sorted heights.

However, both c.blockIDs and c.height are `map`, if they were sorted slice, we could `break` the loop as soon as the height is above the finalized height.